hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java

```diff
@@ -42,14 +44,20 @@ public ContainerInfo getInfo() {
   }
 
   public Set<ContainerReplica> getReplicas() {
-    return new HashSet<>(replicas.values());
+    return replicas;
   }
 
   public ContainerReplica put(ContainerReplica r) {
-    return replicas.put(r.getDatanodeDetails().getID(), r);
+    return copyAndUpdate(map -> map.put(r.getDatanodeDetails().getID(), r));
   }
 
   public ContainerReplica removeReplica(DatanodeID datanodeID) {
-    return replicas.remove(datanodeID);
+    return copyAndUpdate(map -> map.remove(datanodeID));
+  }
+
+  private <T> T copyAndUpdate(java.util.function.Function<Map<DatanodeID, ContainerReplica>, T> update) {
+    T result = update.apply(this.replicaMap);
+    this.replicas = Collections.unmodifiableSet(new HashSet<>(this.replicaMap.values()));
```

Comment:
nit: let's avoid `this.` like the rest of the class

```suggestion
    T result = update.apply(replicaMap);
    replicas = Collections.unmodifiableSet(new HashSet<>(replicaMap.values()));
```